mainchain-follower/db-sync-follower/src/mc_hash/mod.rs

```diff
@@ -0,0 +1,59 @@
+use crate::{block::BlockDataSourceImpl, metrics::McFollowerMetrics, observed_async_trait};
+use async_trait::async_trait;
+use main_chain_follower_api::{common::Timestamp, DataSourceError};
+use sidechain_domain::McBlockHash;
+use sidechain_mc_hash::McHashDataSource;
+use std::sync::Arc;
+
+pub struct McHashDataSourceImpl {
+	inner: Arc<BlockDataSourceImpl>,
+	metrics_opt: Option<McFollowerMetrics>,
+}
+
+impl McHashDataSourceImpl {
+	pub fn new(inner: Arc<BlockDataSourceImpl>, metrics_opt: Option<McFollowerMetrics>) -> Self {
+		Self { inner, metrics_opt }
+	}
+}
+
+observed_async_trait!(
+impl McHashDataSource for McHashDataSourceImpl {
+	type Error = DataSourceError;
+
+	async fn get_latest_stable_block_for(
+		&self,
+		reference_timestamp: sp_timestamp::Timestamp,
+	) -> std::result::Result<Option<sidechain_mc_hash::MainchainBlock>, Self::Error> {
+		Ok(self.inner.get_latest_stable_block_for(
+			Timestamp(reference_timestamp.as_millis())
+		)
+		.await?
+		.map(|block| sidechain_mc_hash::MainchainBlock {
```

Comment:
I would go with `.map(From::from)`. Especially if return type is changed to `domain::MainchainBlock`

Reply:
all those casts went away completely after switching to `domain::MainchainBlock`